### PR TITLE
[Server] WMS Fix the mandatory OnlineResource

### PR DESCRIPTION
**Pull request by sbrunner**
I not sure that well understand the expected result, I see that:

* in the WMS specification the OnlineResource is required in the Service tag.
* in the Service tag we get the OnlineResource from the project configuration.
* in the other tag we automatically generate the OnlineResource from the request.

What should we finally do?

### CODE REVIEW

**Comment by elpaso** [comment]
@sbrunner the expected behavior is:
- if the onlineresource is specfied in the OWS project settings, use that one
- if it is not specified, get it from request

**Comment by sbrunner** [comment]
@elpaso make sense, then I will do ti for all the OnlineResource?

**Review by rldhont — CHANGES_REQUESTED** [review]
@sbrunner the OnlineResource in the Service Tag is not for the Web Service. It's a link for resources about the Web Service.
In the documentation, the OnlineResource in the Service Tag is defined as 
> Top-level web address of service or service provider
In teh QGIS project there is 4 possible OnelineResource:
* 1 for Service description
* 1 for WMS service
* 1 for WFS service
* 1 for WCS service

**Comment by sbrunner** [comment]
@rldhont in a QGIS server all of them is the same addresses (as for mapserver)... And the "Service description" is as I know not an ogc service ...

**Comment by rldhont** [comment]
@sbrunner All the OnlineResource can be different not only because you can defined the forth in QGIS but because some users don't want to use the same address for the different services.

The OnlineResource in the Service Tag is defined for WMS as the _Top-level web address of service or service provider_ and it is stored in the project as* WMSOnlineResource*.

The OnlineResource for WMS service used for request is stored in the project as *WMSUrl*, if it is not defined get the URL from the request.

The OnlineResource for WFS service used for request is stored in the project as *WFSUrl*, if it is not defined get the URL from the request.

The OnlineResource for WCS service used for request is stored in the project as *WCSUrl*, if it is not defined get the URL from the request.

I don't know if all of these rules has been used well implemented.

File: src/server/services/wms/qgswmsgetcapabilities.cpp [hunk 1]
```diff
@@ -283,14 +283,15 @@ namespace QgsWms
...
-      onlineResourceElem.setAttribute( QStringLiteral( "xmlns:xlink" ), QStringLiteral( "http://www.w3.org/1999/xlink" ) );
-      onlineResourceElem.setAttribute( QStringLiteral( "xlink:type" ), QStringLiteral( "simple" ) );
-      onlineResourceElem.setAttribute( QStringLiteral( "xlink:href" ), onlineResource );
-      serviceElem.appendChild( onlineResourceElem );
+      onlineResource = serviceUrl( request, project );
     }
+    QDomElement onlineResourceElem = doc.createElement( QStringLiteral( "OnlineResource" ) );
+    onlineResourceElem.setAttribute( QStringLiteral( "xmlns:xlink" ), QStringLiteral( "http://www.w3.org/1999/xlink" ) );
+    onlineResourceElem.setAttribute( QStringLiteral( "xlink:type" ), QStringLiteral( "simple" ) );
+    onlineResourceElem.setAttribute( QStringLiteral( "xlink:href" ), serviceUrl( request, project ) );
```
Comment (rldhont) [inline src/server/services/wms/qgswmsgetcapabilities.cpp]: This line has to be updated:
```
onlineResourceElem.setAttribute( QStringLiteral( "xlink:href" ), onlineResource );
```

**Review by rldhont — CHANGES_REQUESTED** [review]

**Comment by sbrunner** [comment]
@rldhont thanks for your response, I get a better understanding, but what should be do if the WMSOnlineResource config isn't fill?
May last solution isn't good? Should I change something?

File: src/server/services/wms/qgswmsgetcapabilities.cpp (hunk 1 above)
Comment (sbrunner) [inline src/server/services/wms/qgswmsgetcapabilities.cpp]: thanks :-)

Comment (rldhont) [inline src/server/services/wms/qgswmsgetcapabilities.cpp]: You calculated onlineResource and does not use it in the element.

**Comment by sbrunner** [comment]
@rldhont Corrected and tests pass :-)

**Review by rldhont — APPROVED** [review]

**Comment by rldhont** [comment]
Thanks @sbrunner 
Will you check WMSUrl ?

**Comment by sbrunner** [comment]
It's checked in `serviceUrl` :-)

**Comment by rldhont** [comment]
Thanks, I will check for WFS and WCS.

**Pull request merged**